src/ann_solo/reader.py

```diff
@@ -511,3 +966,162 @@ def read_mztab_ssms(filename: str) -> pd.DataFrame:
+        random_permutation = np.random.permutation(
+            [i for i in range(len(seq_shuffled)) if
+             i not in indices_to_exclude])
+        random_permutation = random_permutation.tolist()
+        full_permutation = [
+            random_permutation.pop(0) if i not in indices_to_exclude else i for
+            i in range(len(seq_shuffled))]
+        seq_shuffled = seq_shuffled[full_permutation]
+        # Compute the similarity between seq_shuffled and seq_originalusing
+        #  edit distance
+        edit_distance = sum(
+            1 for x in ndiff(seq_shuffled.tolist(), seq_original) if
+            x[0] != ' ')
+        similarity = 1 - (edit_distance / len(seq_original))
+        # Check if similarity is below the specified threshold
+        if similarity <= max_similarity:
+            return ''.join(seq_shuffled), {full_permutation[i]: i  for i in
+                                           range(len(seq_original))}
+        elif similarity < best_similarity:
+            best_similarity, best_shuffled, best_permutation = similarity, ''.join(
+                seq_shuffled), full_permutation
+    return best_shuffled, {full_permutation[i]: i  for i in
+                           range(len(seq_original))}
+
+
+## Create annotation mapping for reposition support
+def _shuffle_annotation_mapping(sequence_mapping: Dict[int, int]) -> Dict[str,str]:
+    """
+    Create a mapping datastructure of the new annotations that the
+    repositioned peaks will have.
+
+    Parameters
+    ----------
+    sequence_mapping: Dict[int, int]
+        A dictionary mapping the old postions to the new postions of the
+        amino acids in the sequence.
+
+    Returns
+    -------
+    Dict[str,str]
+        A dictionary mapping the old annotations of peaks with their new
+        annotations based on the shuffled sequence.
+    """
+    annotation_mapping = {}
+    sequence_len = len(sequence_mapping)
+    for i in range(1, sequence_len):
+        old_p, new_p = i - 1, sequence_mapping[i - 1]
+        if new_p == sequence_len - 1:
+            annotation_mapping[f"b{i}"] = None
+            annotation_mapping[f"y{sequence_len-i}"] = None
+        elif old_p == new_p:
+            annotation_mapping[f"b{i}"] = f"b{i}"
+            annotation_mapping[f"y{sequence_len-i}"] = f"y{sequence_len-i}"
+        else:
+            annotation_mapping[f"b{i}"] = f"b{new_p + 1}"
+            annotation_mapping[
+                f"y{sequence_len-i}"] = f"y{sequence_len-(new_p+1)}"
+    return annotation_mapping
+
+
+def shuffle_and_reposition(spectrum: MsmsSpectrum) -> MsmsSpectrum:
+    """
+    Creates a decoy spectrum from a real spectrum.
+
+    Parameters
+    ----------
+    spectrum: MsmsSpectrum
+        Real spectrum.
+
+    Returns
+    -------
+    MsmsSpectrum
+        Decoy spectrum.
+    """
+    # annotate original spectrum
+    spectrum.annotate_proforma(spectrum.peptide, 10, "ppm", "by")
```

Comment:
**todo:** Make the precursor tolerance user-configurable. We'll also want to annotate "a" type ions and "p" precursor ions. Also include relevant [precursor losses](https://spectrum-utils.readthedocs.io/en/latest/api.html#spectrum_utils.spectrum.MsmsSpectrum.annotate_proforma).